pkg/features/feature.go

```diff
@@ -0,0 +1,224 @@
+	obj, err := fp.Features.Get(name, v1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("unable to get feature %s: %v", name, err)
+	}
+
+	if obj.Value != nil {
+		*obj.Value = value
+	} else {
+		obj.Value = newBool(value)
+	}
+
+	_, err = fp.Features.Update(obj)
+	if err != nil {
+		return fmt.Errorf("unable to update feature %s: %v", name, err)
+	}
+
+	return nil
+}
+
+// SetAll iterates a feature map, creates all new features and updates existing features with proper defaults
+func (fp *FeatureProvider) SetAll(featuresMap map[string]*Feature) error {
+	fallback := map[string]*bool{}
+
+	for _, feature := range featuresMap {
+		obj, err := fp.Features.Get(feature.name, v1.GetOptions{})
+		if errors.IsNotFound(err) {
+			// since the feature is not found, it will be created
+			newFeature := &v3.Feature{
+				ObjectMeta: v1.ObjectMeta{
+					Name: feature.name,
+				},
+				Default: feature.def,
+				Value:   feature.value,
+			}
+
+			fallback[newFeature.Name] = newFeature.Value
+
+			_, err := fp.Features.Create(newFeature)
+			if err != nil {
+				return fmt.Errorf("could not create feature %s while setting features: %v", feature.name, err)
+			}
+		} else if err != nil {
+			return fmt.Errorf("unable to get feature %s while setting features: %v", feature.name, err)
+		} else {
+			update := false
+			if obj.Default != feature.def {
+				obj.Default = feature.def
+				update = true
+			}
+			if obj.Value != feature.value {
+				obj.Value = feature.value
+				update = true
+			}
+			if update {
+				_, err := fp.Features.Update(obj)
+				if err != nil {
+					return fmt.Errorf("unable to update feature %s while setting features: %v", feature.name, err)
+				}
+			}
+		}
+	}
+
+	fp.Fallback = fallback
+
+	return nil
+}
+
+// Enabled returns whether given feature is enabled
+func (fp *FeatureProvider) Enabled(name string) bool {
+	obj, err := fp.FeaturesLister.Get("", name)
+	if err != nil {
+		if backup := fp.Fallback[name]; backup != nil {
+			return *fp.Fallback[name]
+		}
+		return false
+	}
+	if obj.Value == nil {
+		return obj.Default
+	}
+
+	return *obj.Value
+}
+
+// SetProvider assigns the given provider as the feature provider
+func SetProvider(p *FeatureProvider) error {
+	if err := p.SetAll(globalGate.features); err != nil {
+		return err
+	}
+	provider = p
+	return nil
+}
+
+// Set assigns value to CRD if provider is available, and sets value in feature gate
+func Set(name string, val bool) error {
+	// if provider is unavailable, default for feature is change and will be assigned once provider is up
+	if provider == nil {
+		if f, ok := globalGate.features[name]; ok {
+			b := val
+			if f.value == nil {
+				f.value = &b
+			} else {
+				*f.value = b
+			}
+		}
+	} else {
+		if err := provider.Set(name, val); err != nil {
+			return err
+		}
+	}
+	return SetGate(name, val)
+}
+
+// SetGate sets the value of feature associate with name in global feature gate
+func SetGate(name string, val bool) error {
```

Comment:
Unexport, use `Set` as part of the API for this package so a user can't muck the globalGate